include/seqan3/alphabet/concept.hpp

```diff
@@ -14,24 +14,668 @@
+    template <typename alph_t>
+    //!\cond
+        requires requires (alph_t a, seqan3::alphabet_char_t<alph_t> r)
+        {
+            { seqan3::assign_char_to(r, a) } -> alph_t;
+            { seqan3::char_is_valid_for<alph_t>(r) } -> bool;
+        }
+    //!\endcond
+    decltype(auto) operator()(seqan3::alphabet_char_t<alph_t> const r, alph_t & a) const
+    {
+        if (!seqan3::char_is_valid_for<alph_t>(r))
+            //TODO: instead of doing a narrowing cast here, add constructor overload for other chars?
+            throw seqan3::invalid_char_assignment{seqan3::detail::get_display_name_v<alph_t>, static_cast<char>(r)};
+
+        return seqan3::assign_char_to(r, a);
+    }
+
+    //!\brief Operator overload for rvalues.
+    template <typename alph_t>
+    //!\cond
+        requires requires (alph_t a, seqan3::alphabet_char_t<alph_t> r)
+        {
+            { seqan3::assign_char_to(r, a) } -> alph_t;
+            { seqan3::char_is_valid_for<alph_t>(r) } -> bool;
+        }
+    //!\endcond
+    auto operator()(seqan3::alphabet_char_t<alph_t> const r, alph_t && a) const
+    {
+        return operator()(r, a); // call above function but return by value
+    }
+};
+
+} // namespace seqan3::detail::adl::only
+
+namespace seqan3
+{
+
+/*!\name Function objects
+ * \{
+ */
+
+/*!\brief Assign a character to an alphabet object, throw if the character is not valid.
+ * \tparam your_type Type of the target object.
+ * \param chr  The character being assigned; must be of the seqan3::alphabet_char_t of the target object.
+ * \param alph The target object; its type must model seqan3::Alphabet.
+ * \returns Reference to `alph` if `alph` was given as lvalue, otherwise a copy.
+ * \throws seqan3::invalid_char_assignment If `seqan3::char_is_valid_for<decltype(alph)>(chr) == false`.
+ * \ingroup alphabet
+
+ * \details
+ *
+ * This is a function object. Invoke it with the parameters specified above.
+ *
+ * Note that this is not a customisation point and it cannot be "overloaded".
+ * It simply invokes seqan3::char_is_valid_for and seqan3::assign_char_to.
+ *
+ * ### Example
+ *
+ * \include test/snippet/alphabet/assign_char_strictly_to.cpp
+ *
+ */
+inline constexpr auto assign_char_strictly_to = detail::adl::only::assign_char_strictly_to_fn{};
+//!\}
+} // namespace seqan3
+
+// ============================================================================
+// alphabet_size_v
+// ============================================================================
+
+namespace seqan3::detail::adl::only
+{
+/*!\brief Functor definition used indirectly by for seqan3::detail::alphabet_size_v.
+ * \tparam alph_t   The type being queried.
+ * \tparam s_alph_t `alph_t` with cvref removed and possibly wrapped in std::type_identity; never user-provide this!
+ * \ingroup alphabet
+ */
+template <typename alph_t,
+          typename s_alph_t = std::conditional_t<std::is_nothrow_default_constructible_v<remove_cvref_t<alph_t>> &&
```

Comment:
what happens if it is wrapped in type_identity. What is the use case for this? Only the value_size static member can then be used right? But is it then not a valid alphabet if it only supports alphabet_size() even if it is no nothrow constructible and not constexpr constructible?

Reply:
> ADL looks also in the associated namespaces of the template parameter(!) so the user can provide overloads almost exactly like before.

😲